code/game/atoms.dm

```diff
@@ -360,40 +360,126 @@
-	if(ishuman(M))
-		var/mob/living/carbon/human/H = M
-		blood_type = H.b_type
-	blood_DNA[M.dna.unique_enzymes] = blood_type
-	return 1
+	var/old_length = blood_DNA.len
+	blood_DNA |= blood_dna
+	if(blood_DNA.len > old_length)
+		return 1//some new blood DNA was added
+
 
-// Only adds blood on the floor -- Skie
-/atom/proc/add_blood_floor(mob/living/carbon/M)
-	return //why the fuck this is at an atom level but only works on simulated turfs I don't know
+//to add blood from a mob onto something, and transfer their dna info
+/atom/proc/add_mob_blood(mob/living/M)
+	var/list/blood_dna = M.get_blood_dna_list()
+	if(!blood_dna)
+		return 0
+	var/list/b_data = M.get_blood_data(M.get_blood_id())
+
+	return add_blood(blood_dna, b_data["blood_color"])
+
+//to add blood onto something, with blood dna info to include.
+/atom/proc/add_blood(list/blood_dna, color)
+	return 0
+
+/obj/add_blood(list/blood_dna, color)
+	return transfer_blood_dna(blood_dna)
+
+/obj/item/add_blood(list/blood_dna, color)
+	var/blood_count = !blood_DNA ? 0 : blood_DNA.len
+	if(!..())
+		return 0
+	if(!blood_count)//apply the blood-splatter overlay if it isn't already in there
+		add_blood_overlay(blood_DNA, color)
+	return 1 //we applied blood to the item
+
+/obj/item/proc/add_blood_overlay(list/blood_dna, color)
+	if(initial(icon) && initial(icon_state))
+		//try to find a pre-processed blood-splatter. otherwise, make a new one
+		var/index = blood_splatter_index()
+		var/icon/blood_splatter_icon = blood_splatter_icons[index]
+		if(!blood_splatter_icon)
+			blood_splatter_icon = icon(initial(icon), initial(icon_state), , 1)		//we only want to apply blood-splatters to the initial icon_state for each object
+			blood_splatter_icon.Blend(color, ICON_ADD) 			//fills the icon_state with white (except where it's transparent)
+			blood_splatter_icon.Blend(icon('icons/effects/blood.dmi', "itemblood"), ICON_MULTIPLY) //adds blood and the remaining white areas become transparant
+			blood_splatter_icon = fcopy_rsc(blood_splatter_icon)
+			blood_splatter_icons[index] = blood_splatter_icon
+		overlays += blood_splatter_icon
+
+/obj/item/clothing/gloves/add_blood(list/blood_dna, color)
+	. = ..()
+	transfer_blood = rand(2, 4)
+
+/turf/add_blood(list/blood_dna, color)
+	var/obj/effect/decal/cleanable/blood/splatter/B = locate() in src
+	if(!B)
+		B = new /obj/effect/decal/cleanable/blood/splatter(src)
+	B.transfer_blood_dna(blood_dna) //give blood info to the blood decal.
+	B.basecolor = color
+	return 1 //we bloodied the floor
+
+/mob/living/carbon/human/add_blood(list/blood_dna, color)
+	if(wear_suit)//i don't think here is appllying colors either.
+		wear_suit.add_blood(blood_dna, color)
+		update_inv_wear_suit()
+	else if(w_uniform)
+		w_uniform.add_blood(blood_dna)
+		update_inv_w_uniform()
+	if(gloves)
+		var/obj/item/clothing/gloves/G = gloves
+		G.add_blood(blood_dna, color)
+	else
+		hand_blood_color = color
+		bloody_hands = rand(2, 4)
+
+	transfer_blood_dna(blood_dna)

```

Comment:
Why move it back? all the other `add_blood`'s above already call this in their own procs; I believe this should only be handled if the `else` condition is reached---unless there's some explicit reason for it.